src/fromager/bootstrapper.py

```diff
@@ -145,6 +196,56 @@ def _processing_build_requirement(self, current_req_type: RequirementType) -> bo
         return False
 
     def bootstrap(self, req: Requirement, req_type: RequirementType) -> Version:
+        """Bootstrap a package and its dependencies.
+
+        In test mode, catches all exceptions, records failures, and continues.
+        In normal mode, raises exceptions immediately (fail-fast).
+        """
+        try:
+            return self._bootstrap_impl(req, req_type)
+        except Exception as err:
+            if not self.test_mode:
+                raise
+            return self._handle_bootstrap_failure(req, err)
+
+    def _handle_bootstrap_failure(self, req: Requirement, err: Exception) -> Version:
+        """Handle complete bootstrap failure in test mode.
+
+        Called when an exception escapes from _bootstrap_impl(). Records the
+        failure and returns a version to allow processing to continue.
+        """
+        logger.error(
+            "test mode: failed to bootstrap %s: %s",
+            req,
+            err,
+            exc_info=True,
+        )
+
+        # Try to get resolved version from cache if available
+        cached = self._resolved_requirements.get(str(req))
+        resolved_version = cached[1] if cached else None
+
+        # Get source type if we have a resolved version
+        if resolved_version:
+            source_url_type = str(sources.get_source_type(self.ctx, req))
+        else:
+            source_url_type = "unknown"
+
+        self.failed_builds.append(
+            BuildFailure.from_exception(
+                req=req,
+                resolved_version=resolved_version,
+                source_url_type=source_url_type,
+                exception=err,
+            )
+        )
+
+        # Return a version to allow processing to continue
+        # Use cached version if available, otherwise return a null version
+        return resolved_version or Version("0.0.0")
```

Comment:
Does anything actually use the version that is returned by `bootstrap()`? I thought when we looked at the code we found that nothing did.